storage/engine/db.cc

```diff
@@ -607,19 +611,61 @@ class DBMergeOperator : public rocksdb::MergeOperator {
 
 class DBLogger : public rocksdb::Logger {
  public:
-  DBLogger(DBLoggerFunc f)
-      : func_(f) {
+  DBLogger(bool enabled)
+      : enabled_(enabled) {
   }
   virtual void Logv(const char* format, va_list ap) {
-    // TODO(pmattis): forward to Go logging. Also need to benchmark
-    // calling Go exported methods from C++ to determine if this is
-    // too slow.
-    vfprintf(stderr, format, ap);
-    fprintf(stderr, "\n");
+    // TODO(pmattis): Benchmark calling Go exported methods from C++
+    // to determine if this is too slow.
+    if (!enabled_) {
+      return;
+    }
+
+    // First try with a small fixed size buffer.
+    char space[1024];
+
+    // It's possible for methods that use a va_list to invalidate the data in
+    // it upon use. The fix is to make a copy of the structure before using it
+    // and use that copy instead.
+    va_list backup_ap;
+    va_copy(backup_ap, ap);
+    int result = vsnprintf(space, sizeof(space), format, backup_ap);
+    va_end(backup_ap);
+
+    if ((result >= 0) && (result < sizeof(space))) {
+      rocksDBLog(space, result);
+      return;
+    }
+
+    // Repeatedly increase buffer size until it fits.
+    int length = sizeof(space);
+    while (true) {
+      if (result < 0) {
+        // Older behavior: just try doubling the buffer size.
```

Comment:
A negative result may be an error other than insufficient space. We should check errno and/or put some upper limit on length. Or do we need to support this older behavior at all? Both posix and c99 require that the snprintf family of functions return the number of bytes that would be required, and glibc has been doing the right thing since version 2.1 in 1999. I think it's fine to just truncate log messages at 1024 bytes in the unlikely event we encounter a system with old-style snprintf.